source/common/router/config_impl.h

```diff
@@ -344,6 +345,83 @@ class RouteTracingImpl : public RouteTracing {
   envoy::type::FractionalPercent overall_sampling_;
 };
 
+/**
+ * Base implementation for path rewriters
+ */
+class PathRewriter {
+public:
+  virtual bool apply() const = 0;
+  virtual std::string rewrite(std::string path, const std::string& matched_path,
+                              bool case_sensitive) const = 0;
+  virtual ~PathRewriter(){};
+};
+
+/**
+ * Path rewrite through prefix replacement
+ */
+class PrefixPathRewriter : public PathRewriter {
+public:
+  PrefixPathRewriter(std::string prefix) : PathRewriter(), prefix_(prefix) {}
+
+  bool apply() const { return !prefix_.empty(); }
+
+  std::string rewrite(std::string path, const std::string& matched_path,
+                      bool case_sensitive) const {
+    ASSERT(case_sensitive ? absl::StartsWith(path, matched_path)
+                          : absl::StartsWithIgnoreCase(path, matched_path));
+    return path.replace(0, matched_path.size(), prefix_);
```

Comment:
When we get here the prefix ends up being the same for each test case in config_impl_test. 

Ex.
```
[2019-10-02 16:40:30.436][14][debug][misc] [bazel-out/k8-fastbuild/bin/source/common/router/_virtual_includes/config_lib/common/router/config_impl.h:399] the prefix found is: /api/new_endpoint
test/common/router/config_impl_test.cc:615: Failure
Expected equality of these values:
  "/bar"
  headers.get_(Http::Headers::get().Path)
    Which is: "/api/new_endpoint"
```